File: dat-nuget/Saro.Dat/DatCertificate.cs
namespace Saro.Dat;

public class DatCertificate : ICloneable, IDisposable
{
    public long Cid { get; }
    internal IDatSignature Signature { get; }
    internal IDatCrypto Crypto { get; }
    public long DatIssuanceStartSeconds { get; }
    public long DatIssuanceEndSeconds { get; }
    public long DatTtlSeconds { get; }
    public long ExpireSeconds { get; }

    private bool _disposed;

    private DatCertificate(long cid, IDatSignature signature, IDatCrypto crypt, long ib, long ie, long datTtlSeconds, long expireSeconds)
    {
        Cid = cid;
        Signature = signature;
        Crypto = crypt;
        DatIssuanceStartSeconds = ib;
        DatIssuanceEndSeconds = ie;
        DatTtlSeconds = datTtlSeconds;
        ExpireSeconds = expireSeconds;
    }

    public bool Expired => ExpireSeconds < Unixtime.Now();
    public bool Issuable
    {
        get
        {
            long now = Unixtime.Now();
            return Signature.Signable() && now >= DatIssuanceStartSeconds && now <= DatIssuanceEndSeconds;
        }
    }

    public string Exports(bool verifyOnly = false)
    {
        return $"{Cid:x}.{DatIssuanceStartSeconds}.{DatIssuanceEndSeconds - DatIssuanceStartSeconds}.{DatTtlSeconds}.{Signature.Algorithm().ToText()}.{Crypto.Algorithm().ToText()}.{DatUtils.EncodeBase64Url(Signature.ExportKey(verifyOnly))}.{DatUtils.EncodeBase64Url(Crypto.ToBytes())}";
    }

    public override string ToString() => Exports(false);

    public object Clone() => new DatCertificate(Cid, Signature.Clone(), Crypto.Clone(), DatIssuanceStartSeconds, DatIssuanceEndSeconds, DatTtlSeconds, ExpireSeconds);

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        (Signature as IDisposable)?.Dispose();
        (Crypto as IDisposable)?.Dispose();
    }

    public override bool Equals(object? obj)
    {
        if (obj is DatCertificate other) return this.Cid == other.Cid;
        return false;
    }

    public override int GetHashCode() => Cid.GetHashCode();

    public static DatCertificate Generate(long cid, long datIssuanceStartSeconds, long datIssuanceDurationSeconds, long datTtlSeconds, DatSignatureAlgorithm sa, DatCryptoAlgorithm ca)
    {
        return New(
            cid,
            datIssuanceStartSeconds,
            datIssuanceDurationSeconds,
            datTtlSeconds,
            IDatSignature.Generate(sa),
            IDatCrypto.Generate(ca)
        );
    }

    public static DatCertificate New(long cid, long datIssuanceStartSeconds, long datIssuanceDurationSeconds, long datTtlSeconds, IDatSignature sk, IDatCrypto ck)
    {
        if (sk is null) throw new DatException(DatErrorCode.ConfigArgumentInvalid, "signature key is null");
        if (ck is null) throw new DatException(DatErrorCode.ConfigArgumentInvalid, "crypto key is null");
        if (datIssuanceStartSeconds < 0) throw new DatException(DatErrorCode.CertMalformed, "datIssuanceStartSeconds must be >= 0");
        if (datIssuanceDurationSeconds < 0) throw new DatException(DatErrorCode.CertMalformed, "datIssuanceDurationSeconds must be >= 0");
        if (datTtlSeconds < 0) throw new DatException(DatErrorCode.CertMalformed, "datTtlSeconds must be >= 0");

        long datIssuanceEndSeconds;
        long expireSeconds;
        try
        {
            checked
            {
                datIssuanceEndSeconds = datIssuanceStartSeconds + datIssuanceDurationSeconds;
                expireSeconds = datIssuanceEndSeconds + datTtlSeconds;
            }
        }
        catch (OverflowException e)
        {
            throw new DatException(DatErrorCode.CertMalformed,
                "datIssuanceStartSeconds + datIssuanceDurationSeconds + datTtlSeconds overflowed", e);
        }

        return new DatCertificate(cid, sk, ck, datIssuanceStartSeconds, datIssuanceEndSeconds, datTtlSeconds, expireSeconds);
    }

    public static DatCertificate Parse(string format)
    {
        if (format is null)
            throw new DatException(DatErrorCode.ConfigArgumentInvalid, "certificate string is null");

        var p = format.Split('.');
        if (p.Length != 8)
            throw new DatException(DatErrorCode.CertMalformed, "expected exactly 8 dot-separated fields");

        IDatSignature? signatureKey = null;
        try
        {
            long cid = Field(DatErrorCode.CertMalformed, "cid field is not a plain hex integer",
                () => DatUtils.ParseCidStrict(p[0]));
            long datIssuanceStartSeconds = Field(DatErrorCode.CertMalformed, "issuance_start_seconds field is not a plain decimal integer",
                () => DatUtils.ParseSecondsStrict(p[1]));
            long datIssuanceDurationSeconds = Field(DatErrorCode.CertMalformed, "issuance_duration_seconds field is not a plain decimal integer",
                () => DatUtils.ParseSecondsStrict(p[2]));
            long datTtlSeconds = Field(DatErrorCode.CertMalformed, "dat_ttl_seconds field is not a plain decimal integer",
                () => DatUtils.ParseSecondsStrict(p[3]));

            var signatureAlgorithm = DatSignatureAlgorithmExtensions.FromText(p[4]);
            var cryptAlgorithm = DatCryptoAlgorithmExtensions.FromText(p[5]);

            byte[] signatureKeyBytes = Field(DatErrorCode.CertMalformed, "signature_key field is not base64url",
                () => DatUtils.DecodeBase64Url(p[6]));
            byte[] cryptoKeyBytes = Field(DatErrorCode.CertMalformed, "crypto_key field is not base64url",
                () => DatUtils.DecodeBase64Url(p[7]));

            signatureKey = IDatSignature.FromKey(signatureAlgorithm, signatureKeyBytes);
            var cryptoKey = IDatCrypto.FromBytes(cryptAlgorithm, cryptoKeyBytes);

            var certificate = New(cid, datIssuanceStartSeconds, datIssuanceDurationSeconds, datTtlSeconds, signatureKey, cryptoKey);
            signatureKey = null;
            return certificate;
        }
        catch
        {
            (signatureKey as IDisposable)?.Dispose();
            throw;
        }
    }

    private static T Field<T>(string code, string detail, Func<T> read)
    {
        try
        {
            return read();
        }
        catch (DatException)
        {
            throw;
        }
        catch (Exception e)
        {
            throw new DatException(code, detail, e);
        }
    }
}

File: dat-maven/src/main/kotlin/me/saro/dat/dat/DatCertificate.kt
package me.saro.dat.dat

import me.saro.dat.DatUtils
import me.saro.dat.Unixtime
import me.saro.dat.crypto.DatCrypto
import me.saro.dat.crypto.DatCryptoAlgorithm
import me.saro.dat.exception.DatErrorCode
import me.saro.dat.exception.DatException
import me.saro.dat.signature.DatSignature
import me.saro.dat.signature.DatSignatureAlgorithm

class DatCertificate private constructor(
    val cid: ULong,
    internal val signature: DatSignature,
    internal val crypto: DatCrypto,
    internal val datIssuanceStartSeconds: ULong,
    internal val datIssuanceEndSeconds: ULong,
    internal val datTtlSeconds: ULong,
): Cloneable {
    internal val cidHex = cid.toString(16)
    internal val cidHexBytes = cidHex.toByteArray()

    val expired: Boolean get() = (datIssuanceEndSeconds + datTtlSeconds) < Unixtime.now().toULong()

    val issuable: Boolean get() {
        return this.signable() && Unixtime.now().toULong() in datIssuanceStartSeconds..datIssuanceEndSeconds
    }

    val cidLong: Long get() = cid.toLong()

    fun signable(): Boolean {
        return signature.signable()
    }
    fun supportVerifyOnly(): Boolean {
        return signature.supportVerifyOnly()
    }

    fun exports(verifyOnly: Boolean = false): String {
        return StringBuilder()
            .append(cidHex).append('.')
            .append(datIssuanceStartSeconds).append('.')
            .append(datIssuanceEndSeconds - datIssuanceStartSeconds).append('.')
            .append(datTtlSeconds).append('.')
            .append(signature.algorithm()).append('.')
            .append(crypto.algorithm()).append('.')
            .append(DatUtils.encodeBase64Url(signature.exportKey(verifyOnly))).append('.')
            .append(DatUtils.encodeBase64Url(crypto.toBytes()))
            .toString()
    }

    override fun toString(): String {
        return exports(false)
    }

    public override fun clone(): DatCertificate {
        return DatCertificate(
            cid,
            signature.clone(),
            crypto.clone(),
            datIssuanceStartSeconds,
            datIssuanceEndSeconds,
            datTtlSeconds
        )
    }

    override fun equals(other: Any?): Boolean {
        if (other is DatCertificate) {
            return other.cid == this.cid
        }
        return false
    }

    override fun hashCode(): Int {
        return cid.hashCode()
    }

    companion object {
        @JvmStatic
        fun generate(cid: Long, datIssuanceStartSeconds: Long, datIssuanceDurationSeconds: Long, datTtlSeconds: Long, signatureAlgorithm: DatSignatureAlgorithm, cryptoAlgorithm: DatCryptoAlgorithm): DatCertificate {
            return new(
                cid,
                datIssuanceStartSeconds,
                datIssuanceDurationSeconds,
                datTtlSeconds,
                DatSignature.generate(signatureAlgorithm),
                DatCrypto.generate(cryptoAlgorithm),
            )
        }

        @JvmStatic
        fun new(cid: Long, datIssuanceStartSeconds: Long, datIssuanceDurationSeconds: Long, datTtlSeconds: Long, signatureKey: DatSignature, cryptoKey: DatCrypto): DatCertificate {
            if (datIssuanceStartSeconds < 0L) {
                throw DatException(DatErrorCode.CONFIG_ARGUMENT_INVALID, "datIssuanceStartSeconds must be >= 0")
            }
            if (datIssuanceDurationSeconds < 0L) {
                throw DatException(DatErrorCode.CONFIG_ARGUMENT_INVALID, "datIssuanceDurationSeconds must be >= 0")
            }
            if (datTtlSeconds < 0L) {
                throw DatException(DatErrorCode.CONFIG_ARGUMENT_INVALID, "datTtlSeconds must be >= 0")
            }
            return new(cid.toULong(), datIssuanceStartSeconds.toULong(), datIssuanceDurationSeconds.toULong(), datTtlSeconds.toULong(), signatureKey, cryptoKey)
        }

        @JvmStatic
        fun new(cid: ULong, datIssuanceStartSeconds: ULong, datIssuanceDurationSeconds: ULong, datTtlSeconds: ULong, signatureKey: DatSignature, cryptoKey: DatCrypto): DatCertificate {
            if (datIssuanceDurationSeconds > ULong.MAX_VALUE - datIssuanceStartSeconds) {
                throw DatException(DatErrorCode.CERT_MALFORMED, "issuance_start_seconds + issuance_duration_seconds overflowed u64")
            }
            val datIssuanceEndSeconds = datIssuanceStartSeconds + datIssuanceDurationSeconds
            if (datTtlSeconds > ULong.MAX_VALUE - datIssuanceEndSeconds) {
                throw DatException(DatErrorCode.CERT_MALFORMED, "issuance_start_seconds + issuance_duration_seconds + dat_ttl_seconds overflowed u64")
            }
            return DatCertificate(
                cid,
                signatureKey,
                cryptoKey,
                datIssuanceStartSeconds,
                datIssuanceEndSeconds,
                datTtlSeconds,
            )
        }

        @JvmStatic
        fun parse(format: String): DatCertificate {
            val parts: List<String> = format.split(".")
            if (parts.size != 8) {
                throw DatException(DatErrorCode.CERT_MALFORMED, "expected exactly 8 dot-separated fields")
            }

            val cid = DatUtils.parseU64HexOrNull(parts[0])
                ?: throw DatException(DatErrorCode.CERT_MALFORMED, "cid field is not a plain hex u64")
            val datIssuanceStartSeconds = DatUtils.parseU64OrNull(parts[1])
                ?: throw DatException(DatErrorCode.CERT_MALFORMED, "issuance_start_seconds field is not a plain decimal u64")
            val datIssuanceDurationSeconds = DatUtils.parseU64OrNull(parts[2])
                ?: throw DatException(DatErrorCode.CERT_MALFORMED, "issuance_duration_seconds field is not a plain decimal u64")
            val datTtlSeconds = DatUtils.parseU64OrNull(parts[3])
                ?: throw DatException(DatErrorCode.CERT_MALFORMED, "dat_ttl_seconds field is not a plain decimal u64")

            val signatureAlgorithm = DatSignatureAlgorithm.fromString(parts[4])
            val cryptAlgorithm = DatCryptoAlgorithm.fromString(parts[5])

            val signatureKeyBytes = try {
                DatUtils.decodeBase64Url(parts[6])
            } catch (e: Exception) {
                throw DatException(DatErrorCode.CERT_MALFORMED, "signature_key field is not base64url", e)
            }
            val cryptoKeyBytes = try {
                DatUtils.decodeBase64Url(parts[7])
            } catch (e: Exception) {
                throw DatException(DatErrorCode.CERT_MALFORMED, "crypto_key field is not base64url", e)
            }

            val signatureKey = DatSignature.fromKey(signatureAlgorithm, signatureKeyBytes)
            val cryptoKey = DatCrypto.fromBytes(cryptAlgorithm, cryptoKeyBytes)

            return new(
                cid,
                datIssuanceStartSeconds,
                datIssuanceDurationSeconds,
                datTtlSeconds,
                signatureKey,
                cryptoKey,
            )
        }
    }
}